triedb/pathdb/flush.go

```diff
@@ -63,3 +65,67 @@ func writeNodes(batch ethdb.Batch, nodes map[common.Hash]map[string]*trienode.No
 	}
 	return total
 }
+
+// writeStates flushes state mutations into the provided database batch as a whole.
+//
+// This function assumes the background generator is already terminated and states
+// before the supplied marker has been correctly generated.
+//
+// TODO(rjl493456442) do we really need this generation marker? The state updates
+// after the marker can also be written and will be fixed by generator later if
+// it's outdated.
+func writeStates(batch ethdb.Batch, genMarker []byte, accountData map[common.Hash][]byte, storageData map[common.Hash]map[common.Hash][]byte, clean *fastcache.Cache) (int, int) {
+	var (
+		accounts int
+		slots    int
+	)
+	for addrHash, blob := range accountData {
+		// Skip any account not covered yet by the snapshot. The account
+		// at generation position (addrHash == genMarker[:common.HashLength])
+		// should be updated.
+		if genMarker != nil && bytes.Compare(addrHash[:], genMarker) > 0 {
+			continue
+		}
+		accounts += 1
+		if len(blob) == 0 {
+			rawdb.DeleteAccountSnapshot(batch, addrHash)
+			if clean != nil {
+				clean.Set(addrHash[:], nil)
+			}
+		} else {
+			rawdb.WriteAccountSnapshot(batch, addrHash, blob)
+			if clean != nil {
+				clean.Set(addrHash[:], blob)
+			}
+		}
+	}
+	for addrHash, storages := range storageData {
```

Comment:
I'm wondering whether it makes a difference in the speed of flushing if we sort the keys before trying to insert them, iirc some databases handle ordered key insertions better than random keys. This way we could also break on the first key that exceeds the genMarker